File: CorrelationService.ts
import _ from 'lodash';
import moment from 'moment';

interface Cycle {
  startDate: string;
  length: number;
}

interface Symptom {
  date: string;
  type: string;
  severity: 'Low' | 'Medium' | 'High';
}

export const CorrelationService = {
  /**
   * Analyzes symptoms to find correlations with cycle phases
   */
  analyzeSymptomPatterns: (cycles: Cycle[], symptoms: Symptom[]) => {
    const insights: string[] = [];

    // Group symptoms by type
    const symptomsByType = _.groupBy(symptoms, 'type');

    Object.keys(symptomsByType).forEach(type => {
      const occurrences = symptomsByType[type];
      
      // We need a minimum amount of data to make a correlation
      if (occurrences.length < 3) return;

      // Track which phase the symptom appears in
      const phaseCounts = {
        menstrual: 0, // Days 1-5
        follicular: 0, // Days 6-14
        luteal: 0, // Days 15-28+
      };

      occurrences.forEach(symptom => {
        // Find which cycle this symptom falls into
        const cycle = cycles.find(c => {
          const start = moment(c.startDate);
          const end = moment(c.startDate).add(c.length, 'days');
          return moment(symptom.date).isBetween(start, end, 'day', '[]');
        });

        if (cycle) {
          const dayOfCycle = moment(symptom.date).diff(moment(cycle.startDate), 'days') + 1;
          
          if (dayOfCycle <= 5) phaseCounts.menstrual++;
          else if (dayOfCycle <= 14) phaseCounts.follicular++;
          else phaseCounts.luteal++;
        }
      });

      // Simple heuristic: if > 60% of occurrences happen in one phase
      const total = occurrences.length;
      if (phaseCounts.menstrual / total > 0.6) {
        insights.push(`Your ${type} is strongly linked to your menstrual phase.`);
      } else if (phaseCounts.luteal / total > 0.6) {
        insights.push(`You tend to experience ${type} during your luteal phase (PMS).`);
      }
    });

    return insights;
  }
};

File: app/src/main/java/com/teraxes/vital/domain/CorrelationService.kt
package com.teraxes.vital.domain

import com.teraxes.vital.data.model.CycleEntity
import com.teraxes.vital.data.model.SymptomLogEntity
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.TimeUnit

object CorrelationService {

    /**
     * Analyzes logged symptoms against recorded cycles to find phase correlations
     */
    fun analyzeSymptomPatterns(
        cycles: List<CycleEntity>,
        symptoms: List<SymptomLogEntity>
    ): List<String> {
        val insights = mutableListOf<String>()
        if (symptoms.isEmpty()) {
            insights.add("Log more symptoms daily to receive personalized pattern insights.")
            return insights
        }

        val dateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
        val symptomsByType = symptoms.groupBy { it.symptomType }

        symptomsByType.forEach { (type, occurrences) ->
            if (occurrences.size < 2) return@forEach

            var menstrualCount = 0
            var follicularCount = 0
            var lutealCount = 0
            var matched = 0

            occurrences.forEach { symptom ->
                val symptomDateMs = try {
                    dateFormat.parse(symptom.date)?.time
                } catch (e: Exception) {
                    null
                } ?: return@forEach

                // Find matching cycle
                val matchedCycle = cycles.find { cycle ->
                    val cycleEnd = cycle.endDate ?: (cycle.startDate + TimeUnit.DAYS.toMillis(35))
                    symptomDateMs in cycle.startDate..cycleEnd
                }

                if (matchedCycle != null) {
                    val dayOfCycle = (TimeUnit.MILLISECONDS.toDays(symptomDateMs - matchedCycle.startDate) + 1).toInt()
                    val phase = CycleUtils.getCyclePhase(dayOfCycle, CycleUtils.getCycleLength(matchedCycle))
                    when (phase) {
                        "menstrual" -> menstrualCount++
                        "follicular" -> follicularCount++
                        "ovulation", "luteal" -> lutealCount++
                    }
                    matched++
                }
            }

            if (matched > 0) {
                val total = matched.toDouble()
                if (menstrualCount / total >= 0.5) {
                    insights.add("Your $type is strongly linked to your menstrual phase.")
                } else if (lutealCount / total >= 0.5) {
                    insights.add("You tend to experience $type during your luteal phase (PMS).")
                } else if (follicularCount / total >= 0.5) {
                    insights.add("Your $type frequently occurs during your follicular phase.")
                }
            }
        }

        if (insights.isEmpty()) {
            insights.add("Keep logging symptoms over multiple cycles to uncover patterns!")
        }

        return insights
    }
}
